### ISSUE TITLE
[BUG] No backup progress indicated

### ISSUE THREAD

**Issue by ChipWolf** [longhorn/longhorn#2785]
**Describe the bug**
We have a few large volumes (100GB).
They have a scheduled daily backup and occasionally we run a backup manually.

As of writing this, there were backups taken 19 hours ago, and further backups 6 hours ago according to the volume's snapshot tree in the Longhorn UI.
![](https://i.cwlf.uk/OmxB3)

However, the latest backup according to the backups page was 18 hours ago.
![](https://i.cwlf.uk/xjaBK)

We observe this behaviour daily; a backup is either initiated manually or automatically. Unpredictably, after several hours, it'll appear in the backups list.
However, counter-intuitively, the backups always appear instantaneously in the snapshot tree.

We don't see an indicator anywhere of the progress of ongoing backups other than it eventually appearing on the backups page.

**To Reproduce**
1. Have a large volume full of actual data.
2. Create a backup
3. Observe the backup appearing in the snapshot tree immediately
4. Observe the backup being made available to restore in the backup page several hours later.

**Expected behavior**
We expect to see some kind of progress indicator or a clearer view in the snapshot tree that the backup is not actually available to restore yet.

**Environment:**
 - Longhorn version: 1.1.2
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: k3s v1.20.6+k3s1
   - Number of management node in the cluster: 1
   - Number of worker node in the cluster: 0
 - Node config
   - OS type and version: Ubuntu 20.04.2
   - CPU per node: 16
   - Memory per node: 64G
   - Disk type(e.g. SSD/NVMe): NVMe
   - Network bandwidth between the nodes: N/A
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): OVH Bare Metal
 - Number of Longhorn volumes in the cluster: 11

**Additional context**
Detailed above is our single-node development "cluster"; however, we see the same behaviour in our 3-node production cluster.

**Comment by ChipWolf** [comment]
Possibly related to #1667, but it doesn't look like an identical issue

**Comment by jenting** [comment]
The blue icon means snapshot; the green icon means backup.
From the image, I can't see the backup.

Could you please check the longhorn-manager Pod's log to see if the HTTP server receive the request of `snapshotBackup`? I'm worried that the longhorn manager did not receive the request to create a snapshot backup.

Also, you could check the `kubectl get engines.longhorn.io -o yaml`, it contains the backup progress under `status.backupStatus.<backup-name>.progress`.

**Comment by ChipWolf** [comment]
Ah, right. In that case, I have not seen a green icon on this volume.

This is the log from longhorn-manager after choosing 'create backup':
```
time="2021-07-09T11:34:18Z" level=debug msg="Created snapshot ebfe16c3-27a5-4d78-ae62-a6bb3def77d4 with labels map[] for volume pvc-25afa2a0-732b-457f-9453-f6b43397d65b"
- - [09/Jul/2021:11:34:18 +0000] "POST /v1/volumes/pvc-25afa2a0-732b-457f-9453-f6b43397d65b?action=snapshotCreate HTTP/1.1" 200 406 "https://longhorn.example.org/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0"
```

There are no backupStatus objects for the volume:
```
    backupStatus: {}
    currentImage: longhornio/longhorn-engine:v1.1.2
    currentReplicaAddressMap:
      pvc-25afa2a0-732b-457f-9453-f6b43397d65b-r-d2c68424: 10.42.0.222:10000
```

**Comment by ChipWolf** [comment]
Watched the logs for a little while longer, it fails to create the backup:

```
time="2021-07-09T11:44:39Z" level=error msg="Failed to initiate backup for snapshot 9b87c0df-3501-4219-85ae-9f2c0868f77d of volume pvc-25afa2a0-732b-457f-9453-f6b43397d65b with label map[KubernetesStatus:{\"pvName\":\"pvc-25afa2a0-732b-457f-9453-f6b43397d65b\",\"pvStatus\":\"Bound\",\"namespace\":\"test\",\"pvcName\":\"test-workload-datadir\",\"lastPVCRefAt\":\"\",\"workloadsStatus\":[{\"podName\":\"test-workload-855bcbbb48-cwkz8\",\"podStatus\":\"Running\",\"workloadName\":\"test-workload-855bcbbb48\",\"workloadType\":\"ReplicaSet\"}],\"lastPodRefAt\":\"\"} type:test]" error="failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.2/longhorn [--url 10.42.0.190:10007 backup create --dest s3://longhorn@us-east-1/ --label type=test --label KubernetesStatus={\"pvName\":\"pvc-25afa2a0-732b-457f-9453-f6b43397d65b\",\"pvStatus\":\"Bound\",\"namespace\":\"test\",\"pvcName\":\"test-workload-datadir\",\"lastPVCRefAt\":\"\",\"workloadsStatus\":[{\"podName\":\"test-workload-855bcbbb48-cwkz8\",\"podStatus\":\"Running\",\"workloadName\":\"test-workload-855bcbbb48\",\"workloadType\":\"ReplicaSet\"}],\"lastPodRefAt\":\"\"} 9b87c0df-3501-4219-85ae-9f2c0868f77d], output , stderr, time=\"2021-07-09T11:41:39Z\" level=info msg=\"Backing up 9b87c0df-3501-4219-85ae-9f2c0868f77d on tcp://10.42.0.222:10000, to s3://longhorn@us-east-1/\"\ntime=\"2021-07-09T11:44:39Z\" level=fatal msg=\"Error running create backup command: failed to create backup to s3://longhorn@us-east-1/ for volume pvc-25afa2a0-732b-457f-9453-f6b43397d65b: rpc error: code = DeadlineExceeded desc = context deadline exceeded\"\n, error exit status 1"
time="2021-07-09T11:44:39Z" level=warning msg="HTTP handling error fail to backup snapshot: failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.2/longhorn [--url 10.42.0.190:10007 backup create --dest s3://longhorn@us-east-1/ --label type=test --label KubernetesStatus={\"pvName\":\"pvc-25afa2a0-732b-457f-9453-f6b43397d65b\",\"pvStatus\":\"Bound\",\"namespace\":\"test\",\"pvcName\":\"test-workload-datadir\",\"lastPVCRefAt\":\"\",\"workloadsStatus\":[{\"podName\":\"test-workload-855bcbbb48-cwkz8\",\"podStatus\":\"Running\",\"workloadName\":\"test-workload-855bcbbb48\",\"workloadType\":\"ReplicaSet\"}],\"lastPodRefAt\":\"\"} 9b87c0df-3501-4219-85ae-9f2c0868f77d], output , stderr, time=\"2021-07-09T11:41:39Z\" level=info msg=\"Backing up 9b87c0df-3501-4219-85ae-9f2c0868f77d on tcp://10.42.0.222:10000, to s3://longhorn@us-east-1/\"\ntime=\"2021-07-09T11:44:39Z\" level=fatal msg=\"Error running create backup command: failed to create backup to s3://longhorn@us-east-1/ for volume pvc-25afa2a0-732b-457f-9453-f6b43397d65b: rpc error: code = DeadlineExceeded desc = context deadline exceeded\"\n, error exit status 1"
time="2021-07-09T11:44:39Z" level=error msg="Error in request: fail to backup snapshot: failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.2/longhorn [--url 10.42.0.190:10007 backup create --dest s3://longhorn@us-east-1/ --label type=test --label KubernetesStatus={\"pvName\":\"pvc-25afa2a0-732b-457f-9453-f6b43397d65b\",\"pvStatus\":\"Bound\",\"namespace\":\"test\",\"pvcName\":\"test-workload-datadir\",\"lastPVCRefAt\":\"\",\"workloadsStatus\":[{\"podName\":\"test-workload-855bcbbb48-cwkz8\",\"podStatus\":\"Running\",\"workloadName\":\"test-workload-855bcbbb48\",\"workloadType\":\"ReplicaSet\"}],\"lastPodRefAt\":\"\"} 9b87c0df-3501-4219-85ae-9f2c0868f77d], output , stderr, time=\"2021-07-09T11:41:39Z\" level=info msg=\"Backing up 9b87c0df-3501-4219-85ae-9f2c0868f77d on tcp://10.42.0.222:10000, to s3://longhorn@us-east-1/\"\ntime=\"2021-07-09T11:44:39Z\" level=fatal msg=\"Error running create backup command: failed to create backup to s3://longhorn@us-east-1/ for volume pvc-25afa2a0-732b-457f-9453-f6b43397d65b: rpc error: code = DeadlineExceeded desc = context deadline exceeded\"\n, error exit status 1"
```

**Comment by PhanLe1010** [comment]
@ChipWolf 
When Longhorn takes a backup of a volume, it first takes a snapshot then creates a backup from that snapshot.

The icons you see in Longhorn UI are snapshots and they will appear almost immediately because (snapshot is an O(1) operation in Longhorn). The snapshot should change from blue to green once the corresponding backup finishes. 

In these cases, they did not change to green, and there are multiple of them at nearly the same time, and there are `initiate backup` errors as you mentioned. Like you said it sound similar to one of the old bugs but you are using v1.1.2 which is supposed to have those bugs fixed. Can you send a support bundle to investigate?

**Comment by ChipWolf** [comment]
I've generated a support bundle @PhanLe1010 it asked for the issue URL so I'm guessing it's uploaded somewhere, or shall I attach it to this issue myself?

**Comment by PhanLe1010** [comment]
Please send it to `longhorn-support-bundle@rancher.com`

**Comment by PhanLe1010** [comment]
Received support bundle over email. Thanks @ChipWolf! 

It seems that replica client timeout while waiting for the volume's replica to start the backup: 
```
# Logs from longhorn manager pod
2021-07-15T12:04:09.632956841Z time="2021-07-15T12:04:09Z" level=warning msg="HTTP handling error fail to backup snapshot: failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.2/longhorn [--url 10.42.0.190:10007 backup create --dest s3://longhorn@us-east-1/ --label KubernetesStatus={\"pvName\":\"pvc-25afa2a0-732b-457f-9453-f6b43397d65b\",\"pvStatus\":\"Bound\",\"namespace\":\"farcemint\",\"pvcName\":\"farcemint-forge-ranchdream-farcemint-datadir\",\"lastPVCRefAt\":\"\",\"workloadsStatus\":[{\"podName\":\"farcemint-forge-ranchdream-farcemint-855bcbbb48-kw5n7\",\"podStatus\":\"Running\",\"workloadName\":\"farcemint-forge-ranchdream-farcemint-855bcbbb48\",\"workloadType\":\"ReplicaSet\"}],\"lastPodRefAt\":\"\"} --label RecurringJob=c-orwcyy c-orwcyy-caed624e], output , stderr, time=\"2021-07-15T12:01:09Z\" level=info msg=\"Backing up c-orwcyy-caed624e on tcp://10.42.0.222:10000, to s3://longhorn@us-east-1/\"\ntime=\"2021-07-15T12:04:09Z\" level=fatal msg=\"Error running create backup command: failed to create backup to s3://longhorn@us-east-1/ for volume pvc-25afa2a0-732b-457f-9453-f6b43397d65b: rpc error: code = DeadlineExceeded desc = context deadline exceeded\"\n, error exit status 1"
```
I am guessing it is the 1-minute `GRPCServiceCommonTimeout` timeout in this replica client function https://github.com/longhorn/longhorn-engine/blob/370c56957758f41a8e8357f63c1f74b17b5ebdcb/pkg/replica/client/client.go#L461-L486. This could happen when this [preload()](https://github.com/longhorn/longhorn-engine/blob/370c56957758f41a8e8357f63c1f74b17b5ebdcb/pkg/replica/backup.go#L304-L306) have to process big snapshots with many extents. I notice that the problematic volume in this issue has actually size of 200GB and ~30 snapshots. 

I think the Longhorn team should try to see if we can reproduce this issue in our lab:

1. Create a 100 GB volume,
2. Write data/create snapshots so that the volume has about 200GB actual size 
3. Perform backup
4. Check if we hit the same timeout error

**Comment by PhanLe1010** [comment]
@ChipWolf  We are aware of this limitation with big volume. We are planning to fix it as part of https://github.com/longhorn/longhorn/issues/2218

**Comment by ChipWolf** [comment]
Thanks, @PhanLe1010; oddly we're not seeing this issue with an identical configuration on our 3-node production cluster
![](https://i.cwlf.uk/P2aBG)

Seems to only be an issue on this single-node cluster. Minio is backed off to the same NFS with both clusters.

**Comment by joshimoo** [comment]
Once ref #1761 #2623 has been added we can transition away from relying on the engine in memory status. That should also allow for adding the ability to see how long a backup took.

**Comment by jenting** [comment]
Related #1311.

**Comment by ChipWolf** [comment]
> Seems to only be an issue on this single-node cluster. Minio is backed off to the same NFS with both clusters.

Scratch this, the first backup succeeded, subsequent backups are doing the same thing
![](https://i.cwlf.uk/aGyGO)

**Comment by shuo-wu** [comment]
It's not related to the cluster node count. The timeout is triggered by the long initialization for the large volume backup. I already plan to track and fix this issue in #1761 and #2807.
